gef.py

```diff
@@ -6821,37 +6816,45 @@ def do_invoke(self, argv):
         return
 
     @staticmethod
-    def get_main_thread_tcache():
-        """The tcache is consistently the first allocation in the main arena,
-        so we can reliably guess the location of the tcache in the main thread.
-        """
-        heap_base = HeapBaseFunction.heap_base()
-        if heap_base is None:
-            err("No heap section")
-            return 0x0
-        return heap_base + 0x10
+    def find_tcache():
+        """Return the location of the current thread's tcache."""
+        try:
+            # For multithreaded binaries, the tcache symbol (in thread local
+            # storage) will give us the correct address.
+            tcache_addr = gdb.parse_and_eval("(void *) tcache")
+        except gdb.error:
+            # In binaries not linked with pthread (and therefore there is only
+            # one thread), we can't use the tcache symbol, but we can guess the
+            # correct address because the tcache is consistently the first
+            # allocation in the main arena.
+            heap_base = HeapBaseFunction.heap_base()
+            if heap_base is None:
+                err("No heap section")
+                return 0x0
+            tcache_addr = heap_base + 0x10
```

Comment:
I think here `2*current_arch.ptrsize` is better than hardcoding `0x10`

Reply:
I think the correct value is `0x10` whether the architecture is 64- or 32-bit. But I'm getting some weird results now that I inspect 32-bit test cases closely. I'll probably be pushing more corrections shortly. :/

Reply:
Confirmed that `0x10` is correct regardless of word size.

But then I went down a rabbit hole relating to the size of each chunk in a tcache bin.

![tcache-incorrect-size](https://user-images.githubusercontent.com/22407135/114790919-930ed000-9d53-11eb-8801-9eaae1bb588a.png)

As you can see in the screenshot, `gef`'s "tcachebin size" doesn't match the "chunk size" when working with a 32-bit binary. The actual pattern is apparent:

**32-bit**
| tcachebin index | chunk size |
--- | --- 
| 0 | 16 |
| 1 | 32 |
| 2 | 48 |

**64-bit**
| tcachebin index | chunk size |
--- | --- 
| 0 | 32 |
| 1 | 48 |
| 2 | 64 |

I could update `gef` to match the pattern, but I've been experimenting and can't come up with a theory as to exactly _why_ the pattern exists. 

The [source](https://elixir.bootlin.com/glibc/glibc-2.29/source/malloc/malloc.c#L311) provides a macro for converting a chunksize to a tcachebin index, but I can't make my understanding of the source match what I'm seeing experimentally. For example, my understanding is that a chunk that's 48 bytes should be stored in tcachebin 4 in a 32-bit environment.<sup>1</sup> But the screenshot shows it stored in tcachebin 2.

So I can definitely commit some code that calculates `(idx * 16) + (ptrsize*4)` to get the chunk size from the bin index, but I can't point to glibc source code that says that's so, which makes me nervous. So I guess I'll keep searching for another day or so, before I give up and say that experimental data beats my apparently incorrect understanding of the source.

[1]: MINSIZE = 16; ALIGNMENT = 8; (48 - 16 + 8 -1) / 8 = 39 / 8 = 4

Reply:
Hmm, this is weird indeed. How about setting the tcachebin size to be `chunk.get_chunk_size()`? In that case, we don't need to worry too much about whether the pattern always holds, and the command output still be showing the correct size.

Reply:
That's a hell of a description @irontigran thank you for that. It is a weird behavior and I could confirm it too on an old x86 I had around with libc 2.27.

Since this behavior doesn't really come in the scope of what this PR is about, I think it's okay to merge this PR and continue the discussion if we need to change something. Thoughts @Grazfather @daniellimws ?

Reply:
Yup we can merge first.